# Computing Galois Groups

The goal of this assignment is to get some experience using SageMath to compute Galois groups of polynomials.  

### Using built-in methods
Of course, there are built-in methods to do so.  (They end up using pari.)  You should be able to guess the name of the function to call on a polynomial using tab completion.

In [ ]:
S.<x> = PolynomialRing(QQ)
f = x^7 + 20*x + 16

In [ ]:
G = f.galois_group()

There is a built-in list of transitive subgroups of small degree, so the answer references this list.  Given the group, you can ask about the order, generators, and so forth.  You can use the built-in method to check your answers, but for the rest of the assignment we want to explore doing so ourselves.

In [ ]:
groups7 = sage.groups.perm_gps.permgroup_named.TransitiveGroups(7)

print(f"There are {len(groups7)} transitive groups of degree 7.")

for G in groups7:
    print("Order",G.order(),"and generators",G.gens())

## Galois Groups for Polynomials of Degree 7

I have given you a bunch of polynomials of degree 7.  By reducing modulo p, find the likely cycle shapes of elements and guess which group it is.  Write up your guesses and a brief justification in the markdown cell "Answer 1".

Hints: polynomials have a change_ring method you will like

In [ ]:
S.<x> = PolynomialRing(QQ)
f1 = x^7 + 20*x + 16
f2 = x^7 + 7 * x^3 + 7 * x^2 + 7*x-1
f3 = x^7+2
f4 = x^7 - 7 * x^3 + 14 *x^2 - 7 *x + 1
f5 = x^7 + 7 * x^4 + 14 *x + 3
f6 = x^7 + x^6 - 12 * x^5 - 7 * x^4 + 28 * x^3 + 14*x^2 - 9 * x + 1
f7 = x^7 - 14 * x^5 + 56*x^3 -56*x + 22

fs = [f1,f2,f3,f4,f5,f6,f7]

In [198]:
def frobshapes(f, n):
    d = f.disc()
    shapes_list = []
    for p in primes(3,n):
        if gcd(p,d) == 1:
            R = Integers(p)
            g = f.change_ring(R)
            G = list(g.factor())
            shape = []
            for i in range(len(G)):
                for j in range(G[i][1]):
                    shape.append(G[i][0].degree())
            shapes_list.append(shape)
    shapes = set(tuple(elem) for elem in shapes_list)
    return(shapes)

i = 1
for f in fs:
    print("f%i"%i)
    print(frobshapes(f, 10000))
    i = i + 1

for G in groups7:
    #print("Order",G.order(),"and generators",G.gens())
    print(G.cycle_index())

f1
{(1, 1, 1, 4), (1, 1, 1, 1, 1, 2), (1, 1, 1, 1, 3), (1, 6), (1, 1, 2, 3), (1, 1, 5), (1, 2, 4), (7,), (2, 2, 3), (2, 5), (1, 1, 1, 2, 2), (3, 4), (1, 2, 2, 2), (1, 3, 3)}
f2
{(7,), (1, 2, 2, 2), (1, 1, 1, 1, 1, 1, 1)}
f3
{(1, 1, 1, 1, 1, 1, 1), (1, 6), (1, 2, 2, 2), (7,), (1, 3, 3)}
f4
{(1, 1, 1, 1, 1, 1, 1), (1, 2, 4), (1, 1, 1, 2, 2), (7,), (1, 3, 3)}
f5
{(1, 1, 1, 1, 3), (1, 2, 4), (1, 1, 5), (7,), (1, 1, 1, 2, 2), (2, 2, 3), (1, 3, 3)}
f6
{(7,), (1, 1, 1, 1, 1, 1, 1)}
f7
{(7,), (1, 3, 3), (1, 1, 1, 1, 1, 1, 1)}
1/7*p[1, 1, 1, 1, 1, 1, 1] + 6/7*p[7]
1/14*p[1, 1, 1, 1, 1, 1, 1] + 1/2*p[2, 2, 2, 1] + 3/7*p[7]
1/21*p[1, 1, 1, 1, 1, 1, 1] + 2/3*p[3, 3, 1] + 2/7*p[7]
1/42*p[1, 1, 1, 1, 1, 1, 1] + 1/6*p[2, 2, 2, 1] + 1/3*p[3, 3, 1] + 1/3*p[6, 1] + 1/7*p[7]
1/168*p[1, 1, 1, 1, 1, 1, 1] + 1/8*p[2, 2, 1, 1, 1] + 1/3*p[3, 3, 1] + 1/4*p[4, 2, 1] + 2/7*p[7]
1/2520*p[1, 1, 1, 1, 1, 1, 1] + 1/24*p[2, 2, 1, 1, 1] + 1/36*p[3, 1, 1, 1, 1] + 1/12*p[3, 2, 2] + 1/9*p[3, 3, 1] + 1/4*p[4, 2, 1] + 1/5*

## Answer 1 (10 points)

Gal(f1) is $S_7$

Gal(f2) contains the subgroup of order 14 ($D_7$)

Gal(f3) contains the subgroup of order 42 ($F_7$)

Gal(f4) contains the subgroup of order 168 ($GL_3(\mathbb{F}_2)$)

Gal(f5) contains the subgroup of order 2520 ($A_7$)

Gal(f6) contains the subgroup of order 7 ($C_7$)

Gal(f7) contains the subgroup of order 21 ($C_7 \rtimes C_3$)

## Alternating Group

For those transitive subgroups which you expect to be subgroups of $A_7$, compute the discriminant and check.  To set you up for later, compute approximations to the roots and the discriminant.  Sage can also compute the discriminant to double check your work. Write up your answer to which polynomials have Galois groups inside $A_7$ below.

In [199]:
def my_disc(f):
    roots = f.roots(CC)
    disc = 1
    print(roots)
    for i in range(len(roots)):
        for j in range(i):
            disc = disc*(roots[i][0] - roots[j][0])
    return disc^2
            
#print(my_disc(f1))
#print(f1.disc())

for f in fs:
    print(f.disc())

-73536438796288
-29417779503
-52706752
1666027489
4202539929
171903939769
18078415936


## Answer 2 (10 points)

As you can see above, the discriminants of the first three polynomials are not squares in $\mathbb{Q}$, hence the Galois group of their splitting field does not lie in $A_7$. The rest have squares in $\mathbb{Q}$ as their discriminant, so the Galois group of their splitting field lies in $A_7$.

Hence $Gal(f_5) \cong A_7$.

## Resolvents

Use resolvents to verify some of your guesses as to the Galois groups.  

- One of the Galois groups has order 168 and is isomorphic to $\mathrm{PSL}_2(\mathbf{F}_7)$: check this.  

- One of the Galois groups has order 42 and is isomorphic to a semidirect product of $C_7$ and $C_6$.  Check this one too.

Note: as this is not a course on numerical analysis, don't worry about justifying how precise the results of your computations are.  Simply assume SageMath has computed with enough precision that you can round approximations of integers to the nearest integer.

P. S.  Check out https://people.maths.bris.ac.uk/~matyd/GroupNames/T15.html to see "natural" descriptions of the transitive subgroups

In [239]:
def invariant(H):
    n = H.degree()
    H_list = H.list()
    A = PolynomialRing(QQ, ['x%s'%n for n in range(1,n+1)])
    A_gens = A.gens()
    f = A(0)
    for i in range(len(H_list)):
        h = A(1)
        for j in range(1,n+1):
            #print(H_list[i])
            h = h*(A_gens[H_list[i](j)-1])^(j)
            #print(h)
        f = f + h
        #print(f)
    return(f)

def resolvent(G,H,f):
    rootlist = f.roots(CC)
    roots = []
    for i in range(len(rootlist)):
        roots.append(rootlist[i][0])
    h = invariant(H)
    Q = G.cosets(H, side='right')
    A.<x> = PolynomialRing(CC)
    R = A(1)
    for i in range(len(Q)):
        R = R*(x - h(Q[i][0](roots)))
    return(R)

def resolvent_roots(G,H,f):
    rootslist = resolvent(G,H,f).roots()
    for i in range(len(rootslist)):
        print(rootslist[i][0])
    return 0

#G = groups7[7]; H = groups7[4]; f = f2
# Hence Gal(f2) contained in some conjugate of F_7
#G = groups7[4]; H = groups7[2]; f = f2
# Hence Gal(f2) is contained in some conjugate of D7

#G = groups7[7]; H = groups7[4]; f = f3
# Hence Gal(f4) isomorphic to some conjugate of F_7 

#G = groups7[6]; H = groups7[5]; f = f4
# Hence Gal(f4) isomorphic to some conjugate of GL(3,2)

#G = groups7[6]; H = groups7[3]; f = f7
# Hence Gal(f4) isomorphic to some conjugate of C_7 \rtimes C_3

#G = groups7[6]; H = groups7[5]; f = f6
# Hence Gal(f6) contained in some conjugate of GL(3,2)
#G = groups7[5]; H = groups7[3]; f = f6
# Hence Gal(f6) contained in some conjugate of C_7 \rtimes C_3
#G = groups7[3]; H = groups7[1]; f = f6
# Hence Gal(f6) isomorphic to some conjugate of C_7


#resolvent_roots(G,H,f)

Transitive group number 3 of degree 7

## Answer 3 (20 points)

Briefly explain why your computations verify that your guess as the Galois group is correct.  

$Gal(f_1)$ has enough cycle types to be all of $S_7$.

$Gal(f_5)$ contains $A_5$ by an enumeration of some of its cycle types; it is contained in $A_7$ because its discriminant is a perfect square.

For the other polynomials, we have founded transitive subgroups that are contained in its Galois group and have approximated and factored resolvants to show that each Galois group is contained in the group that we had previously shown it to contain.

$Gal(f_1) \cong S_7$

$Gal(f_2) \cong D_7$

$Gal(f_3) \cong F_7$

$Gal(f_4) \cong GL(2,3)$

$Gal(f_5) \cong A_5$

$Gal(f_6) \cong C_7$

$Gal(f_7) \cong C_7 \rtimes C_3$

## Puzzles

Answer at least one of these if you can.  (5 bonus points each)

- You should be able to understand the Galois group of f6 by pure thought, at least once you know what the answer should be.  How do you do so?

- You should be able to understand the Galois group of f3 by pure thought.  How do you so?

- We discussed degree 5 polynomials in class, so the next natural case would be degree 6.  However, I was nice and gave you degree 7 polynomials in this assignment.  Why are degree 6 polynomials less nice?

- Pick a bunch of random polynomials of degree 7 and look at the relative frequency of different Galois groups (use SageMath's built-in commands as they're more effiicent).  How do your observations fit in with what we discussed in class? It appears that (unless there's something wrong with the below code) that given a random polynomial that happens to be irreducible, its Galois group tends to be all of S_7. 


In [288]:
def tally(n):
    S.<x> = PolynomialRing(QQ)
    total = 0
    groups = [0,0,0,0,0,0,0]
    for i in range(n):
        f = S.random_element(7)
        
        if f.is_irreducible() == True:
            #print(repr(f.galois_group()))
            for j in range(1,8):
                if "number %i"%j in repr(f.galois_group()):
                    groups[j-1] = groups[j-1] + 1
    return groups
print(tally(10000))

[0, 0, 0, 0, 0, 0, 7947]


x^7 + 1/4*x^6 + x^5 - x^4 + x^3 - 704*x^2


In [296]:
f3.galois_group().gens()

[(1,2,3,4,5,6,7), (1,3,2,6,4,5)]